### ISSUE TITLE
Remove ceph-0 node from Ceph storage cluster

### ISSUE THREAD

**Issue by alexander-petkov** [alexander-petkov/wfas#15]
Figure out how to remove and add a storage node to the Ceph cluster, while the rest of the cluster is operating. 

The SATA controller card died, and needs to be replaced. 

I have until Aug 30 to order a new card. The Adaptec card I have chosen has SFF 8643  sockets, so I will need new cables too.

**Comment by alexander-petkov** [comment]
Ceph notes on the Collab site:

https://collab.firelab.org/software/projects/internet2/wiki/Ceph_cluster_set_up

**Comment by alexander-petkov** [comment]
**Steps taken**

1. Disable osd scrubbing:
`sudo ceph osd set noscrub`
`sudo ceph osd set nodeep-scrub`

2. Limit back-fill and recovery:
`ceph tell osd.\* injectargs '--osd-max-backfills=1'`
`ceph tell osd.\* injectargs '--osd-recovery-max-active=1'`
`ceph tell osd.\* injectargs '--osd-recovery-op-priority=1'`

3. List nodes and their OSDs, this is a snipet for ceph-0:
```
ceph node ls osd
{
    "ceph-0": [
        0,
        1,
        3,
        5,
        7,
        10,
        13,
        15,
        18,
        20,
        23,
        26,
        28,
        31,
        33,
        35,
        38,
        40,
        43,
        45,
        48,
        50,
        53,
        56
    ],
```
4. List OSDs that are down:
`ceph osd df|grep down`

5. OSD that are already down should be out, but I did it anyway:
```
for osd in `ceph osd df|grep down|cut -d ' ' -f 1`;do ceph osd out ${osd};done
```
6. Stopping / Removing OSDS
Stop the _down_ OSDs:
```
for osd in `ceph osd df|grep down|cut -d ' ' -f 1`;do systemctl stop ceph-osd@${osd};done
```
7. Stop and disable the remaining OSDs one at time.

`sudo systemctl disable ceph-osd@id`
--Monitor until rebalancing is complete:
`sudo ceph -w`
--Once rebalancing is finished, stop the osd service:
`sudo systemctl stop ceph-osd@id`

**Comment by alexander-petkov** [comment]
To see OSD mapping to a disk device:

`ceph-volume lvm list`

**Comment by alexander-petkov** [comment]
Solutions offered to an identical problem:
https://spinics.net/lists/ceph-users/msg49934.html

1. 

> I usually do ceph osd set noout first to prevent any recoveries
Then replace the hardware and make sure all OSDs come back online
Then ceph osd unset noout

2. 

> I always set "mon_osd_down_out_subtree_limit = host" (and monitor things
of course) since I reckon a down node can often be brought back way faster
than a full rebalance.

**Comment by alexander-petkov** [comment]
I have decided to currently leave ceph-0 with a few OSD's up and running, until it's time to replace hardware. 

Useful links for ceph node maintenance:

Suse:
https://www.suse.com/documentation/ses-4/book_storage_admin/data/storage_bp_inst_cephdeploy_usage.html

Bryce Rowe's wiki notes:
https://collab.firelab.org/software/projects/internet2/wiki/Ceph_upgrade_osds_guide

Ceph docs: adding/removing OSDs:
https://docs.ceph.com/docs/luminous/rados/operations/add-or-rm-osds/#removing-osds-manual

Redhat docs:
https://access.redhat.com/documentation/en-us/red_hat_ceph_storage/2/html/configuration_guide/configuration_reference#general_configuration_reference

https://access.redhat.com/documentation/en-us/red_hat_ceph_storage/2/html/administration_guide/managing_cluster_size

**Comment by alexander-petkov** [comment]
**Removing and adding OSDs**

I followed the Ceph documentation on [Adding/Removing OSDs](https://docs.ceph.com/docs/master/rados/operations/add-or-rm-osds/?highlight=remove), with the addition that I had to remove the volume group from the disk after the destroy command:

```bash
vgremove  ceph-c0ae13a0-4234-4077-9e23-a315b9f9a12b
```
Then follow the rest of the steps in the documentation.

**Comment by alexander-petkov** [comment]
Check default values for these settings and restore them. 
>     2. Limit back-fill and recovery:
>        `ceph tell osd.\* injectargs '--osd-max-backfills=1'`
>        `ceph tell osd.\* injectargs '--osd-recovery-max-active=1'`
>        `ceph tell osd.\* injectargs '--osd-recovery-op-priority=1'`

Then close this ticket
